pocs/filterwheel/libefw.py

```diff
@@ -0,0 +1,281 @@
+import ctypes
+import enum
+import threading
+import time
+
+from pocs.camera.sdk import AbstractSDKDriver
+from pocs.utils import error
+from pocs.utils.library import load_library
+from pocs.utils import CountdownTimer
+
+
+class EFWDriver(AbstractSDKDriver):
+    # Because ZWO EFW library isn't linked properly have to manually load libudev
+    # in global mode first, otherwise get undefined symbol errors.
+    _libudev = load_library('udev', mode=ctypes.RTLD_GLOBAL)
+
+    def __init__(self, library_path=None, **kwargs):
+        """Main class representing the ZWO EFW library interface.
+
+        On construction loads the shared object/dynamically linked version of the EFW SDK library,
+        which must be already installed (see https://astronomy-imaging-camera.com/software-drivers).
+
+        The name and location of the shared library can be manually specified with the library_path
+        argument, otherwise the ctypes.util.find_library function will be used to try to locate it.
+
+        Args:
+            library_path (str, optional): path to the library e.g. '/usr/local/lib/libEFWFilter.so'
+
+        Returns:
+            `~pocs.filter.libefw.EFWDriver`
+
+        Raises:
+            pocs.utils.error.NotFound: raised if library_path not given & find_library fails to
+                locate the library.
+            OSError: raises if the ctypes.CDLL loader cannot load the library.
+        """
+        super().__init__(name='EFWFilter', library_path=library_path, **kwargs)
+
+    # Methods
+
+    def get_SDK_version(self):
+        """Get the version for the SDK."""
+        # EFW SDK has no way to access SDK version.
+        return "Unknown"
+
+    def get_devices(self):
+        """Get connected device 'UIDs' and corresponding device nodes/handles/IDs.
+
+        EFW SDK has no way to access any unique identifier for connected filterwheels.
+        Instead we construct an ID from combination of filterwheel name, number of
+        positions and integer ID. This will probably not be deterministic, in general,
+        and is only guaranteed to be unique between multiple filterwheels on a single
+        computer.
+        """
+        n_filterwheels = self.get_num()  # Nothing works if you don't call this first.
+        filterwheels = {}
+        for i in range(n_filterwheels):
+            fw_id = self.get_ID(i)
+            self.open(fw_id)
+            info = self.get_property(fw_id)
+            self.close(fw_id)
+            filterwheels[f"{info['name']}_{info['slot_num']}_{fw_id}"] = fw_id
+        return filterwheels
```

Comment:
Some amount of exception handling seems warranted. And debug logging.

Reply:
A fair bit of debug logging happens in this method from the other methods that get called, but I've added some checks, error logging, etc.